### PR TITLE
[Impeller] Test `FenceWaiterVK` and fix termination bugs

### PR DESCRIPTION
**Pull request by matanlurey**
Fixes https://github.com/flutter/flutter/issues/134751, and @jonahwilliams suspects it could be related to a number of other flaky/texture leak scenarios (https://github.com/flutter/flutter/issues/133506#issuecomment-1720275749) that only happen sometimes (i.e. on CI but not real devices), i.e. stuff like:

```txt
--- Vulkan Debug Report  ----------------------------------------
|                Severity: Error
|                    Type: { Validation }
|                 ID Name: VUID-vkDestroyBuffer-buffer-00922
|               ID Number: -464217071
|       Queue Breadcrumbs: [NONE]
|  CMD Buffer Breadcrumbs: [NONE]
|         Related Objects: Device [94498356231456] [ImpellerDevice]
|                 Trigger: Validation Error: [ VUID-vkDestroyBuffer-buffer-00922 ] Object 0: handle = 0x55f21cf47d20, name = ImpellerDevice, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0xe4549c11 | Cannot call vkDestroyBuffer on VkBuffer 0xbb00000000bb[] that is currently in use by a command buffer. The Vulkan spec states: All submitted commands that refer to buffer, either directly or via a VkBufferView, must have completed execution (https://www.khronos.org/registry/vulkan/specs/1.3-extensions/html/vkspec.html#VUID-vkDestroyBuffer-buffer-00922)
-----------------------------------------------------------------
```

---

~~This PR will look a bit like a mess until the last 2 PRs merge in, but locally it appears to fix fence races/segfaults that I was seeing on CI, including on Linux with validations enabled. We can test it further tomorrow.~~ EDIT: Updated.

### CODE REVIEW

File: impeller/renderer/backend/vulkan/fence_waiter_vk.cc
Comment (jonahwilliams) [inline impeller/renderer/backend/vulkan/fence_waiter_vk.cc:61]: return false if terminate is true?

File: impeller/renderer/backend/vulkan/context_vk.cc
Comment (jonahwilliams) [inline impeller/renderer/backend/vulkan/context_vk.cc:379]: You also need to add fence waiter and resource manager teardown into the shutdown method, right? otherwise the dtor order is still incorrect

File: impeller/renderer/backend/vulkan/fence_waiter_vk.cc
Comment (matanlurey) [reply impeller/renderer/backend/vulkan/fence_waiter_vk.cc:61]: I think we might not actually want to, at least not terminate_. Imagine the following:

1. Get texture/buffer etc
2. Terminate
3. AddFence

In this case, Vulkan still expects the buffer to be live (and there might be other buffers still in the waiter thread), so we should still be queueing the fence. One thing we could do is return if the thread is _actually_ terminated (joined), which isn't the exact same thing?

Comment (matanlurey) [reply impeller/renderer/backend/vulkan/fence_waiter_vk.cc:61]: See `StillDestroysFenceIfTerminating` in the unit test suite as well

File: impeller/renderer/backend/vulkan/context_vk.cc
Comment (matanlurey) [reply impeller/renderer/backend/vulkan/context_vk.cc:379]: Yeah. I still planned to do this in https://github.com/flutter/engine/pull/45654, I could do it here but I was trying to minimize changes to "what passes on device and CI".

File: impeller/renderer/backend/vulkan/fence_waiter_vk.cc
Comment (jonahwilliams) [reply impeller/renderer/backend/vulkan/fence_waiter_vk.cc:61]: we'll only ever call Terminate or AddFence from the same thread. Once you join, this thread will be blocked on the workload being performed by the fence waiter. So in theory, once Terminate has returned then you'll have already joined the threads - and this should be invalid to add a new fence?

Comment (jonahwilliams) [reply impeller/renderer/backend/vulkan/fence_waiter_vk.cc:61]: So either terminate being true or waiter thread being joined, but make sure you don't let new fences get added and return false or else we're going to have some potentially confusing errors to deal with down the line

File: impeller/renderer/backend/vulkan/context_vk.cc
Comment (jonahwilliams) [reply impeller/renderer/backend/vulkan/context_vk.cc:379]: I think you should do it here and then try to get CI enabled with validation

File: impeller/renderer/backend/vulkan/fence_waiter_vk.cc
Comment (matanlurey) [reply impeller/renderer/backend/vulkan/fence_waiter_vk.cc]: My mistake. Tweaked the code and unit test.

File: impeller/renderer/backend/vulkan/context_vk.cc
Comment (jonahwilliams) [reply impeller/renderer/backend/vulkan/context_vk.cc:379]: That is, we should fix the known errors before we try to land the larger performance/feature changes?

Comment (matanlurey) [reply impeller/renderer/backend/vulkan/context_vk.cc:379]: I would consider this PR fixing a known error, not a performance/feature change but I have no problem adding the Shutdown fix as well - the only reason I was hesitant is without the ResourceManager change I couldn't observe a problem with shutdown (it's also possible this fence waiter PR "makes stuff work").

That being said I have no problem adding the 2 lines to context_vk.cc in this PR. Wdut?

Comment (matanlurey) [reply impeller/renderer/backend/vulkan/context_vk.cc:379]: Ah you mean the resource manager being a performance improvement, that's fair.

File: impeller/renderer/backend/vulkan/test/mock_vulkan.cc
Comment (gaaclarke) [inline impeller/renderer/backend/vulkan/test/mock_vulkan.cc:64]: I think this is unused.

File: impeller/renderer/backend/vulkan/context_vk.cc
Comment (jonahwilliams) [reply impeller/renderer/backend/vulkan/context_vk.cc:379]: If you want, you could patch in the change to add validation and see if it passes without that fix.

Comment (matanlurey) [reply impeller/renderer/backend/vulkan/context_vk.cc:379]: Yeah let me do that locally at least. Good suggestion!

**Review by gaaclarke — APPROVED** [review]
This code is tricky, it looks good to me and the additional tests are an welcome addition.  You should get 2 approvals just because it's tricky threading changes.

One minor nit across the board is that `auto` should only be used if if makes the code easier to read, not out of convenience as per the style guide.  We violate this a lot.

File: impeller/renderer/backend/vulkan/fence_waiter_vk.cc
Comment (matanlurey) [reply impeller/renderer/backend/vulkan/fence_waiter_vk.cc]: We should re-enable the lint! Also fixed.

File: impeller/renderer/backend/vulkan/fence_waiter_vk.cc
Comment (matanlurey) [reply impeller/renderer/backend/vulkan/fence_waiter_vk.cc]: Done!

File: impeller/renderer/backend/vulkan/test/mock_vulkan.cc
Comment (matanlurey) [reply impeller/renderer/backend/vulkan/test/mock_vulkan.cc]: Ah good catch, artifact of old merge. Removed!

(I wish we had private code unused warnings)

File: impeller/renderer/backend/vulkan/fence_waiter_vk.cc
Comment (jonahwilliams) [inline impeller/renderer/backend/vulkan/fence_waiter_vk.cc:113]: Maybe add a note that this doesn't lock because its only called when the threads are joined?

Comment (jonahwilliams) [reply impeller/renderer/backend/vulkan/fence_waiter_vk.cc:113]: Or does it?

File: impeller/renderer/backend/vulkan/fence_waiter_vk.cc
Comment (matanlurey) [reply impeller/renderer/backend/vulkan/fence_waiter_vk.cc:171]: I'll add a TODO here, not because I don't agree but there is a higher chance I'll make a mistake than improve anything and this unblocks several other folks/initiatives.

File: impeller/renderer/backend/vulkan/context_vk.cc
Comment (matanlurey) [reply impeller/renderer/backend/vulkan/context_vk.cc:379]: At least locally, I don't seem to need any other changes.

As discussed on chat, given this is already a risky change, I'm trying not to introduce more chances at changing behavior - happy to still include the termination fixes in a follow-up PR (w/ or w/o the ResourceManager optimizations)

File: impeller/renderer/backend/vulkan/fence_waiter_vk.cc
Comment (matanlurey) [reply impeller/renderer/backend/vulkan/fence_waiter_vk.cc:116]: It doesn't lock because (as-per your suggestion on AddFence), it's not possible to mutate wait_set_ once terminate_ is set.

**Review by jonahwilliams — APPROVED** [review]
LGTM

**Comment by matanlurey** [comment]
@chinmaygarde I have 2 LGTMs, but would prefer you as a third if you're around.

File: impeller/renderer/backend/vulkan/fence_waiter_vk.cc
Comment (jason-simmons) [inline impeller/renderer/backend/vulkan/fence_waiter_vk.cc:61]: Check `terminate_` in the `lock(wait_set_mutex_)` block.

The class should maintain the invariant that `terminate_` is only accessed while holding `wait_set_mutex_`

File: impeller/renderer/backend/vulkan/fence_waiter_vk.cc
Comment (jason-simmons) [inline impeller/renderer/backend/vulkan/fence_waiter_vk.cc:100]: Also do the read of `terminate_` while holding `wait_set_mutex_` here.

It may be easier to reason about this code if a scope is used to control where the `wait_set_mutex_` is held instead of calling `lock.unlock()`.

File: impeller/renderer/backend/vulkan/fence_waiter_vk.cc
Comment (jason-simmons) [inline impeller/renderer/backend/vulkan/fence_waiter_vk.cc:116]: This line is passing a reference to the `wait_set_` member to `Wait`.  So `Wait`'s `wait_set` argument will be an alias for the `wait_set_` member.

`Wait` mutates both `wait_set` and `wait_set_`, which will make this error prone.

This could be simplified by removing the `wait_set` parameter and having `Wait` do the snapshotting of `wait_set_` internally.

File: impeller/renderer/backend/vulkan/fence_waiter_vk.cc
Comment (matanlurey) [reply impeller/renderer/backend/vulkan/fence_waiter_vk.cc]: Thanks! Done.

File: impeller/renderer/backend/vulkan/fence_waiter_vk.cc
Comment (matanlurey) [reply impeller/renderer/backend/vulkan/fence_waiter_vk.cc]: Sounds reasonable. Done.

File: impeller/renderer/backend/vulkan/fence_waiter_vk.cc
Comment (matanlurey) [reply impeller/renderer/backend/vulkan/fence_waiter_vk.cc]: Nice suggestion, I agree the previous code was very complicated. Done.

File: impeller/renderer/backend/vulkan/context_vk.cc
Comment (chinmaygarde) [inline impeller/renderer/backend/vulkan/context_vk.cc:380]: An invalid fence waiter will no longer invalidate the context now. In the future, if fence waiters cannot be created reliably, the call sites of the users will have to be patched.

File: impeller/renderer/backend/vulkan/fence_waiter_vk.cc
Comment (chinmaygarde) [inline impeller/renderer/backend/vulkan/fence_waiter_vk.cc:93]: I have an uninitialized vars phobia. Can you default it to `false` even if its redundant?

File: impeller/renderer/backend/vulkan/fence_waiter_vk.cc
Comment (chinmaygarde) [inline impeller/renderer/backend/vulkan/fence_waiter_vk.cc:65]: The locking without static analysis makes me nervous. I suspect we didn't use the `IPLR_GUARDED_BY` and friends because we don't have static thread safety analysis around condition variable. This should be easy enough to add support for. Can we file a followup bug for it?

File: impeller/renderer/backend/vulkan/fence_waiter_vk.cc
Comment (chinmaygarde) [inline impeller/renderer/backend/vulkan/fence_waiter_vk.cc:122]: Ditto about not having static thread safety analysis because of condition variables. The locking rules should ideally be static checks instead of comments.

**Review by chinmaygarde — APPROVED** [review]
Not having static thread analysis annotated synchronization utils for condition variables really hurts the reviewability of this PR. We should patch that in `//impeller/base/thread.h`. Rather not have to reason about things the compiler can :/

Other than that, just nits.

File: impeller/renderer/backend/vulkan/context_vk.cc
Comment (matanlurey) [reply impeller/renderer/backend/vulkan/context_vk.cc:380]: There is no such thing as an invalid fence waiter, that is, there is no state that ever marks the fence waiter as invalid. I found it misleading that we check a boolean that is effectively constant true - let's it add it back if/when there is such a case?

File: impeller/renderer/backend/vulkan/fence_waiter_vk.cc
Comment (matanlurey) [reply impeller/renderer/backend/vulkan/fence_waiter_vk.cc:65]: Sure, filed https://github.com/flutter/flutter/issues/134843.

Fwiw, this might be yet another place where absl fixes many of these problems for us by having this low-level infra support we don't have to maintain (https://github.com/flutter/flutter/issues/134741).

File: impeller/renderer/backend/vulkan/fence_waiter_vk.cc
Comment (matanlurey) [reply impeller/renderer/backend/vulkan/fence_waiter_vk.cc]: I would say that's consistent with https://google.github.io/styleguide/cppguide.html#Local_Variables as well. Done.

File: impeller/renderer/backend/vulkan/fence_waiter_vk.cc
Comment (matanlurey) [reply impeller/renderer/backend/vulkan/fence_waiter_vk.cc]: Done.

File: impeller/renderer/backend/vulkan/fence_waiter_vk.cc
Comment (matanlurey) [reply impeller/renderer/backend/vulkan/fence_waiter_vk.cc:118]: Ack.

**Review by jason-simmons — APPROVED** [review]

**Pull request merged**
